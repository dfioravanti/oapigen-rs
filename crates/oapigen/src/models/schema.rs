use proc_macro2::TokenStream;
use quote::{quote, ToTokens};
use std::hash::Hash;
use std::{collections::HashMap, fmt};

/// Collects all the imports that were already added to the code.
/// It is used to determine if an import should be added or not.
/// For example assume that two different structs need
/// chrono::DateTime then we add "chrono::DateTime" to the Imports
/// for the first struct that we hit and we will not add it in the second case since we already have it.
pub type Imports = HashMap<String, TokenStream>;

#[derive(Debug)]
pub enum CurrentType {
    Type,
    Const,
    Enum,
    Vec,
    Struct,
}

/// SchemaAsRust represents a rust type that was inferred from a OpenAPI specification.
/// It contains all the information needed to render the type correctly.
#[derive(Debug)]
pub struct SchemaAsRust {
    /// The name of the field as a TokenStream
    pub tokenized_name: TokenStream,
    /// The type represented as a TokenStream
    pub tokenized_type: TokenStream,
    /// The imports needed to make the type compile.
    pub imports: Imports,

    pub current_type: CurrentType,
}
impl Eq for SchemaAsRust {}
impl PartialEq for SchemaAsRust {
    fn eq(&self, other: &Self) -> bool {
        self.to_token_stream().to_string() == other.to_token_stream().to_string()
    }
}

impl Hash for SchemaAsRust {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.to_token_stream().to_string().hash(state);
    }
}

impl fmt::Display for SchemaAsRust {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let imports = self
            .imports
            .iter()
            .fold(TokenStream::new(), |mut acc, (_, v)| {
                acc.extend(v.clone());
                acc
            });

        if imports.is_empty() {
            write!(f, "{}", self.to_token_stream())
        } else {
            write!(f, "{} \n {}", imports, self.to_token_stream())
        }
    }
}

impl ToTokens for SchemaAsRust {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let generated_tokens = match self.current_type {
            CurrentType::Type => tokenize_type(self),
            CurrentType::Const => todo!(),
            CurrentType::Enum => todo!(),
            CurrentType::Vec => todo!(),
            CurrentType::Struct => todo!(),
        };

        tokens.clone_from(&generated_tokens);
    }
}

fn tokenize_type(tokenized_schema: &SchemaAsRust) -> TokenStream {
    let tokenized_name = &tokenized_schema.tokenized_name;
    let tokenized_type = &tokenized_schema.tokenized_type;

    quote! { struct #tokenized_name(#tokenized_type); }
}

#[cfg(test)]
mod tests {
    use super::*;

    use insta::assert_yaml_snapshot;

    #[test]
    fn test_parsed_schema_display() {
        let tokenized_schema = quote! {
            DateTime
        };

        let mut imports = HashMap::new();
        imports.insert(
            "chrono::DateTime".to_string(),
            quote! { use chrono::DateTime; },
        );

        let parsed_schema = SchemaAsRust {
            tokenized_name: quote! { time },
            tokenized_type: tokenized_schema,
            imports,
            current_type: CurrentType::Type,
        };

        let formatted_parsed_schema = format!("{}", parsed_schema);

        assert_yaml_snapshot!(formatted_parsed_schema);
    }
}
