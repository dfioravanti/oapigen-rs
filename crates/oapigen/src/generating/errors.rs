use crate::parsing::errors::ParsingError;
use std::io;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum GeneratingError {
    #[error("Generated rust code cannot be parsed to string.")]
    SynError(#[from] syn::Error),
    #[error("Could not parse spec.")]
    ParseSpecError(#[from] ParsingError),
    #[error("could not write to file.")]
    WriteToFileError(#[from] io::Error),
}
