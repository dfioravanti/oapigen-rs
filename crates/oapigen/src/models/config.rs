use std::path::PathBuf;

pub enum DateTimeLibraries {
    Chrono,
    Jiff,
}

pub struct Libraries {
    pub datetime: DateTimeLibraries,
}

impl Default for Libraries {
    fn default() -> Self {
        Libraries {
            datetime: DateTimeLibraries::Chrono,
        }
    }
}

#[derive(Default)]
pub struct Config {
    pub output_path: PathBuf,
    pub libraries: Libraries,
}
