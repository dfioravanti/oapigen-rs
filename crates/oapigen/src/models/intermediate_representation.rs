#[derive(Debug, serde::Serialize)]
pub struct Struct {
    pub name: String,
    pub description: Option<String>,
    pub fields: Vec<Field>,
}

#[derive(Debug, serde::Serialize)]
pub struct Field {
    pub name: String,
    pub description: Option<String>,
    pub rust_type: String,
    pub is_required: bool,
}

pub struct ConstantField {
    pub name: String,
    pub description: Option<String>,
    pub rust_type: String,
    pub value: serde_json::Value,
}
