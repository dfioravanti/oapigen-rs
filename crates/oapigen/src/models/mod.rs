mod config;
pub mod schema;

pub use config::*;
pub use schema::*;
