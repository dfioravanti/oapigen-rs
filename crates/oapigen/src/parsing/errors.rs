use oas3::spec;
use proc_macro2::LexError;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum ParsingError {
    #[error("could not parse the string to TokenStream.")]
    CannotParseString(#[from] LexError),
    #[error("could not resolve the reference with the passed spec.")]
    ReferenceNotFound(#[from] spec::RefError),
}
