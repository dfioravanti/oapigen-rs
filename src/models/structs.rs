use proc_macro2::TokenStream;
use quote::{ToTokens, TokenStreamExt, quote};
use serde::{Deserialize, Serialize};
use std::{collections::HashMap, fmt};

/// Collects all the imports that were already added to the code.
/// It is used to determine if an import should be added or not.
/// For example assume that two different structs need
/// chrono::DateTime then we add "chrono::DateTime" to the Imports
/// for the first struct that we hit and we will not add it in the second case since we already have it.
pub type Imports = HashMap<String, TokenStream>;

#[derive(Debug)]
pub enum CurrentType {
    Type,
    Const,
    Enum,
    Vec,
    Struct,
}

/// TokenizedType represents a rust type that was inferred from a OpenAPI specification.
/// It contains all the information needed to render the type correctly.
#[derive(Debug)]
pub struct TokenizedSchema {
    /// The name of the field as a TokenStream
    pub tokenized_name: TokenStream,
    /// The type represented as a TokenStream
    pub tokenized_type: TokenStream,
    /// The imports needed to make the type compile.
    pub imports: Imports,

    pub current_type: CurrentType,
}

impl fmt::Display for TokenizedSchema {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let imports = self
            .imports
            .iter()
            .map(|(_, v)| v.to_string())
            .collect::<Vec<_>>();

        write!(
            f,
            "imports => {:?} \n schema => {:?}",
            imports,
            self.to_token_stream().to_string()
        )
    }
}

impl quote::ToTokens for TokenizedSchema {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let tokenized_name = &self.tokenized_name;
        let tokenized_type = &self.tokenized_type;

        let generated_tokens = quote! {

            #tokenized_name: #tokenized_type

        };

        tokens.clone_from(&generated_tokens);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use insta::assert_yaml_snapshot;

    #[test]
    fn test_parsed_schema_display() {
        let tokenized_schema = quote! {
            DateTime
        };

        let mut imports = HashMap::new();
        imports.insert(
            "chrono::DateTime".to_string(),
            quote! { use chrono::DateTime },
        );

        let parsed_schema = TokenizedSchema {
            tokenized_name: quote! { time },
            tokenized_type: tokenized_schema,
            imports,
        };

        let formatted_parsed_schema = format!("{}", parsed_schema);

        assert_yaml_snapshot!(formatted_parsed_schema);
    }
}
