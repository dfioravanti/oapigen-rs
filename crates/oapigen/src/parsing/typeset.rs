use crate::parsing::errors::ParsingError;
use crate::{format, models};
use oas3::spec::{ObjectSchema, SchemaType, SchemaTypeSet};
use proc_macro2::TokenStream;

pub(crate) fn tokenize_schema(
    config: &models::Config,
    schema_name: String,
    schema: ObjectSchema,
) -> Result<models::SchemaAsRust, ParsingError> {
    if schema.const_value.is_some() {}

    let tokenized_schema = match &schema.schema_type {
        Some(schema_typeset) => match schema_typeset {
            SchemaTypeSet::Single(schema_type) => {
                tokenize_type(config, schema_name, &schema, schema_type)
            }
            SchemaTypeSet::Multiple(_items) => todo!(),
        },
        None => todo!(),
    };

    tokenized_schema
}

fn tokenize_type(
    config: &models::Config,
    schema_name: String,
    schema: &ObjectSchema,
    schema_type: &SchemaType,
) -> Result<models::SchemaAsRust, ParsingError> {
    match schema_type {
        SchemaType::Boolean => todo!(),
        SchemaType::Integer => tokenize_integer_schema(schema_name, schema),
        SchemaType::Number => tokenize_number_schema(schema_name, schema),
        SchemaType::String => tokenize_string_schema(config, schema_name, schema),
        SchemaType::Array => todo!(),
        SchemaType::Object => todo!(),
        SchemaType::Null => todo!(),
    }
}

fn tokenize_integer_schema(
    schema_name: String,
    schema: &ObjectSchema,
) -> Result<models::SchemaAsRust, ParsingError> {
    // if one injects schema_name directly into quote it gets tokenized together with "",
    // which we do not want. Doing it like this drops the ""
    let tokenized_name: TokenStream = schema_name.parse()?;

    let (tokenized_type, imports) = match &schema.format {
        Some(format) => format::format_number(format),
        None => format::format_number(format::DEFAULT_INTEGER),
    };

    Ok(models::SchemaAsRust {
        tokenized_name,
        tokenized_type,
        imports,
        current_type: models::CurrentType::Type,
    })
}

fn tokenize_number_schema(
    schema_name: String,
    schema: &ObjectSchema,
) -> Result<models::SchemaAsRust, ParsingError> {
    // if one injects schema_name directly into quote it gets tokenized together with "",
    // which we do not want. Doing it like this drops the ""
    let tokenized_name: TokenStream = schema_name.parse()?;

    let (tokenized_type, imports) = match &schema.format {
        Some(format) => format::format_number(format),
        None => format::format_number(format::DEFAULT_NUMBER),
    };

    Ok(models::SchemaAsRust {
        tokenized_name,
        tokenized_type,
        imports,
        current_type: models::CurrentType::Type,
    })
}

fn tokenize_string_schema(
    config: &models::Config,
    schema_name: String,
    schema: &ObjectSchema,
) -> Result<models::SchemaAsRust, ParsingError> {
    // if one injects schema_name directly into quote it gets tokenized together with "",
    // which we do not want. Doing it like this drops the ""
    let tokenized_name: TokenStream = schema_name.parse()?;

    let (tokenized_type, imports) = match &schema.format {
        Some(format) => format::format_string(config, format),
        None => format::format_string(config, format::DEFAULT_STRING),
    };

    Ok(models::SchemaAsRust {
        tokenized_name,
        tokenized_type,
        imports,
        current_type: models::CurrentType::Type,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    use crate::models::{DateTimeLibraries, Libraries};
    use rstest::rstest;

    #[rstest]
    #[case("integer", "Age", "type: integer")]
    #[case("number", "Height", "type: number")]
    #[case("string", "Name", "type: string")]
    fn test_parse_base_cases(
        #[case] name: &str,
        #[case] schema_name: &str,
        #[case] schema_spec: &str,
    ) {
        let config = models::Config {
            output_path: Default::default(),
            libraries: Libraries {
                datetime: DateTimeLibraries::Chrono,
            },
        };

        let mut insta_settings = insta::Settings::clone_current();
        insta_settings.set_snapshot_suffix(name);

        let schema = serde_yaml::from_str::<ObjectSchema>(schema_spec).unwrap();

        let got = tokenize_schema(&config, schema_name.to_string(), schema).unwrap();

        insta_settings.bind(|| {
            insta::assert_yaml_snapshot!(got.to_string());
        });
    }

    /*

    TODO: fix constant test
       #[rstest]
       #[case("integer", "Age", "const: 3", "type Age = i32 ;")]
       fn test_parse_constants(
           #[case] name: &str,
           #[case] schema_name: &str,
           #[case] schema_spec: &str,
           #[case] expected: &str,
       ) {
           let mut settings = insta::Settings::clone_current();
           settings.set_snapshot_suffix(name);

           let schema = serde_yaml::from_str::<ObjectSchema>(schema_spec).unwrap();

           let got = tokenize_schema(schema_name.to_string(), schema);

           settings.bind(|| {
               insta::assert_yaml_snapshot!(got.to_string());
           });
           assert_eq!(got.to_string(), expected.to_string());
       }

    */
}
