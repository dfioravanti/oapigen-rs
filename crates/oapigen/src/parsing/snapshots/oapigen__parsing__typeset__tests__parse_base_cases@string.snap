---
source: crates/oapigen/src/parsing/typeset.rs
expression: got.to_string()
---
struct Name (String) ;
